text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,464 @@
+---
+feature name: cli-framework-compatibility-strategy
+start date: February 2nd, 2020
+rfc pr: https://github.com/aws/aws-cdk-rfcs/pull/111
+related issue: https://github.com/aws/aws-cdk-rfcs/issues/110
+---
+
+<!-- replace the blockquoted sections with your content -->
+# Summary
+
+This RFC addresses compatibility concerns for developing different parts of the CDK.
+
+It details exactly what compatibility promises we should be advertising to our customers,
+as well as proposes technical mechanisms for validating those promises are upheld.
+
+At a high level, we define the expected customer experience when upgrading either of the two CDK components:
+
+- CLI upgrades are compatible. New CLI versions should work with older framework versions, and all existing functionality is preserved.
+- Framework upgrades are incompatible. We will require the user to upgrade the CLI as well.
+
+This will eventually lead the user to always have a setup where `CLI >= Framework`.
+
+# Motivation
+
+The motivation behind this RFC is three-fold:
+
+## Confusing Upgrade Process
+
+Currently, the upgrade process for our users can be a bit complicated/confusing.
+
+We generally don't require users to upgrade their CLI when they upgrade the framework.
+
+However, sometimes, we introduce features in the framework that rely on new CLI capabilities.
+When this happens, we bump the cx-protocol version, forcing (all) users to upgrade their CLI as well, even those who don't use the new feature.
+
+Things get much worse, if we forget to bump the cx-protocol version (which we do sometimes).
+In this case, the user will not receive the upgrade instruction, and will rightfully assume the new feature simply works.
+While in reality, this can either cause unexpected errors or non-working features.
+
+We want to come up with a consistent model that users can rely on.
+
+## Simplify CX Protocol Versioning
+
+Recently, we have had a few customer impacting issues that relate to our cx-protocol.
+
+- [aws/aws-cdk: A newer version of the CDK CLI (>= 1.21.0) is necessary to interact with this app](https://github.com/aws/aws-cdk/issues/5986)
+- [aws/aws-cdk: CDK CLI can only be used with apps created by CDK >= 1.10.0](https://github.com/aws/aws-cdk/issues/4294)
+
+A major contributing factor to those issues, is that the ergonomics of introducing breaking changes (cx-protocol version bump), are not very good.
+Something about the process feels convoluted and we seem to be getting it wrong almost every time we attempt it.
```

Comment:
> feels convoluted

That's not a very compelling argument. Generally speaking, would like to have clearer indication of what's convoluted... So we can make sure we actually fix it with the proposed solution.

Reply:
I'll see if i can phrase something more coherent. Though i don't think it's worth our time actually doing a full analysis on the current state, since we are introducing pretty massive changes. 

The most important thing is to know where we want to be, and make sure these mechanisms get us there.

Reply:
I agree with @iliapolo - since this RFC proposes to get rid of this mechanism altogether, It's not worth spending too much energy on trying to identify where the current approach is broken. Suffice it to say that we almost *never* got it right.